### DISCUSSION TITLE
key pair connection with snowflake

### DISCUSSION THREAD

**Discussion by ericfreitas3** [trinodb/trino#24286] [Ideas]
In snowflake recents upgrade, for security, it was included a new feature that classifies users into personal access or service access, users with personal access are forced to use MFA and users with service access are forced to use a RSA key pair connection. 

Until then i've been trying to find a way to connect trino to snowflake using service access (key pair), but the connector doesnt give that option and i didnt find any PR about it, today im using a legacy definition of user in snowflake because of that, there is another way? I think it will be a great advance for snowflake users who work with trino.

**Comment by hashhar** [comment]
cc: @dprophet

**Comment by ebyhr** [comment]
You can use `connection-url` as the workaround. I tried it a few month ago and it worked well. 
Note that the value will be printed in bootstrap.

**Reply by bluck7** [reply]
Can you provide an example url with private key path and passphrase?

**Reply by ericfreitas3** [reply]
i've used like this: connection-url=jdbc:snowflake://XXXXXXXXXXXXX&private_key_path=XXXXXXXXXX i've tested without passphrase but i wil try with it.

**Reply by bluck7** [reply]
Thank you! I finally got it to work using private_key_path and private_key_pwd

**Reply by ericfreitas3** [reply]
your key format is p8?

**Reply by hpopuri2** [reply]
@ericfreitas3  what we have to do if our file is in p8 format ?

**Comment by ericfreitas3** [comment]
i've tried to use the connection url but im getting this error:

Query 20241206_174018_00134_28n39 failed: Connection string is invalid. Unable to parse.

i've used like this: connection-url=jdbc:snowflake://XXXXXXXXXXXXX&private_key_path=XXXXXXXXXX

**Reply by rponsaerts** [reply]
Indeed, it does work well via the connection-url. You were almost there :)

I'm using Snowflake jdbc driver 3.20.0.

E.g.Try 
`connection-url=jdbc:snowflake://<account>.snowflakecomputing.com/?private_key_file=/etc/trino/my_privkey.crt`

Have a look at the "Private key file name and password in connection string" section at
[https://docs.snowflake.com/en/developer-guide/jdbc/jdbc-configure](https://docs.snowflake.com/en/developer-guide/jdbc/jdbc-configure)

`jdbc:snowflake://<account>.snowflakecomputing.com/?private_key_file=/tmp/rsa_key.p8&private_key_file_pwd=dummyPassword`

**Reply by ericfreitas3** [reply]
it almost worked, now im getting this error:

failed: Private key provided is invalid or not supported

tried to convert to .PEM but also doesnt work.

thats my url now: jdbc:snowflake://XXXXXXXXX.snowflakecomputing.com/?private_key_file=XXXXXXXXXXXXX&private_key_pwd=XXXXXXXXXXXXXX

**Reply by rponsaerts** [reply]
perhaps try first with the decrypted private key.

I assume you have the other parameters as well in your snowflake.properties catalog file:

`connector.name=snowflake`
`connection-url=jdbc:snowflake://<account>.snowflakecomputing.com/?private_key_file=/etc/trino/my_privkey.crt`
`connection-user=<user>`
`snowflake.account=<account>`
`snowflake.role=<role>`
`snowflake.database=<database>`
`snowflake.warehouse=<warehouse>`

**Reply by rponsaerts** [reply]
Also, looks like you need to replace private_key_pwd in the url by private_key_**file**_pwd.

**Reply by ericfreitas3** [reply]
connector.name=snowflake
connection-url=jdbc:snowflake://XXXXXXXXXXXX.snowflakecomputing.com/?private_key_file=rsa_key_service.p8&private_key_file_pwd=XXXXXXXXXXXX
connection-user=XXXXXXXXXXXX
connection-password=XXXXXXXXXXXX
snowflake.account=XXXXXXXXXXXX
snowflake.database=XXXXXXXXXXXX
snowflake.role=XXXXXXXXXXXX
snowflake.warehouse=XXXXXXXXXXXX

this is the structure of my catalog (adjusted after the last comment), im getting this error: 

failed: Private key provided is invalid or not supported

i'm using this key in another ingestion proccess (snowpark), also generated a .PEM version of it and it gives the same error.

**Reply by harsh-im** [reply]
were you able to resolve this @ericfreitas3 ?

**Comment by TrungHuynhPB** [comment]
I got it to work as I have to create a new key (.p8) without encrypted password.

Adding Snowflake Bouncy Castle options to jvm.config:
-Dnet.snowflake.jdbc.enableBouncyCastle=TRUE
-Dorg.bouncycastle.asn1.allow_unsafe_integer=true

and also in docker container settings for Trino:
    environment:
      - _JAVA_OPTIONS=-Dfile.encoding=UTF-8 -Dnet.snowflake.jdbc.enableBouncyCastle=TRUE -Dorg.bouncycastle.asn1.allow_unsafe_integer=true

My .properties:
connector.name=snowflake
connection-url=jdbc:snowflake://<account>.snowflakecomputing.com/?user=<user>&warehouse=<WH_name>&db=<DB-name>&schema=<schema-name>&private_key_file=/secrets/rsa_key.p8
connection-user=user
connection-password=na #not use it but still keeps the var here
snowflake.account=account
snowflake.database=DB_name
snowflake.role=RL_name
snowflake.warehouse=WH_name

**Comment by YechielLeader** [comment]
Is this the recommended way to Deploy Snowflake integration nowadays?  Legacy is going to drop soon

**Comment by ebyhr** [comment]
https://github.com/trinodb/trino/pull/26519 is adding support for RSA private key authentication.
